src/fabric/test/fabric2_doc_size_tests.erl

```diff
@@ -0,0 +1,340 @@
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(fabric2_doc_size_tests).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+
+% Doc body size calculations
+% ID: size(Doc#doc.id)
+% Rev: size(erlfdb_tuple:encode(Start)) + size(Rev) % where Rev is usually 16
+% Deleted: 1 % (binary value is one byte)
+% Body: couch_ejson_size:external_size(Body) % Where empty is {} which is 2)
+
+
+-define(NUM_RANDOM_TESTS, 1000).
+
+
+-define(DOC_IDS, [
+    {0, <<>>},
+    {1, <<"a">>},
+    {3, <<"foo">>},
+    {6, <<"foobar">>},
+    {32, <<"af196ae095631b020eedf8f69303e336">>}
+]).
+
+-define(REV_STARTS, [
+    {1, 0},
+    {2, 1},
+    {2, 255},
+    {3, 256},
+    {3, 65535},
+    {4, 65536},
+    {4, 16777215},
+    {5, 16777216},
+    {5, 4294967295},
+    {6, 4294967296},
+    {6, 1099511627775},
+    {7, 1099511627776},
+    {7, 281474976710655},
+    {8, 281474976710656},
+    {8, 72057594037927935},
+    {9, 72057594037927936},
+    {9, 18446744073709551615},
+
+    % The jump from 9 to 11 bytes is because when we
+    % spill over into the bigint range of 9-255
+    % bytes we have an extra byte that encodes the
+    % length of the bigint.
+    {11, 18446744073709551616}
+]).
+
+-define(REVS, [
+    {0, <<>>},
+    {8, <<"foobarba">>},
+    {16, <<"foobarbazbambang">>}
+]).
+
+-define(DELETED, [
+    {1, true},
+    {1, false}
+]).
+
+-define(BODIES, [
+    {2, {[]}},
+    {13, {[{<<"foo">>, <<"bar">>}]}},
+    {28, {[{<<"b">>, <<"a">>}, {<<"c">>, [true, null, []]}]}}
+]).
+
+-define(ATT_NAMES, [
+    {5, <<"a.txt">>},
+    {7, <<"foo.csv">>},
+    {29, <<"a-longer-name-for-example.bat">>}
+]).
+
+-define(ATT_TYPES, [
+    {24, <<"application/octet-stream">>},
+    {10, <<"text/plain">>},
+    {9, <<"image/png">>}
+]).
+
+-define(ATT_BODIES, [
+    {0, <<>>},
+    {1, <<"g">>},
+    {6, <<"foobar">>},
+    {384, <<
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+        "xlasdjfsapoiewrposdlfadfuaducvwerwlkdsfljdfusfsd"
+    >>}
+]).
+
+-define(ATT_HEADERS, [
+    {0, undefined},
+    {2, {[]}},
+    {13, {[{<<"foo">>, <<"bar">>}]}},
+    {32, {[{<<"a">>, <<"header">>}, {<<"b">>, <<"such header">>}]}}
+]).
+
+-define(LDOC_IDS, [
+    {7, <<"_local/">>},
```

Comment:
This is technically invalid 

```
 $ http put $DB1/db/_local/ x=y

{
    "error": "bad_request",
    "reason": "Invalid _local document id."
}
```

Reply:
Totes. Saw that when I got to testing the document id validation thing the other day. Will remove it to avoid confusion even though its not asserted at this level.